File: online/uva-oj/495_Fibo_Freeze.cpp
/***************************************************************************************************
* Title : Fibonacci Freeze
* URL   : http://uva.onlinejudge.org/external/5/495.html
* Comp  :
* Status: Accepted (time 0.029)
* Notes : 
* Ack   : Md Abdul Kader(http://dal.cs.utep.edu/members/mkader/) for BigInteger Class
* meta  : tag-algo-dp, tag-big-integer
***************************************************************************************************/
#include <cstdio>
#include <cstring>
using namespace std;

// ref: 'ds/BigInt.cpp'
class BigInt;

int main() {
  int i, n;

  BigInt num[5001];
  num[0] = BigInt("0");
  num[1] = BigInt("1");

  for (i = 2; i < 5001; i++)
    num[i] = num[i - 1] + num[i - 2];

  while (scanf("%d", &n) != EOF) {
    printf("The Fibonacci number for %d is ", n);
    num[n].show();
  }

  return 0;
}

File: online/uva-oj/495_Fibo_Freeze.java
/***************************************************************************************************
* Title : Fibonacci Freeze
* URL   : 495
* Status: Accpeted
* Author: Atiq Rahman
* meta: tag-big-integer
***************************************************************************************************/
import java.util.Scanner;
import java.math.BigInteger;

class Main {
  public static void main (String args[]) {
    Scanner istream = new Scanner(System.in);
    BigInteger fibo[] = new BigInteger[5001];
    fibo[0] = new BigInteger("0");
    fibo[1] = new BigInteger("1");
    int i, n;
    
    for (i=2; i<5001; i++) {
      fibo[i] = new BigInteger("0");
      fibo[i] = fibo[i].add(fibo[i-2]);
      fibo[i] = fibo[i].add(fibo[i-1]);
    }
    while (istream.hasNextInt()) {
      n = istream.nextInt();
      System.out.println("The Fibonacci number for "+n+" is "+fibo[n]);
    }
  }
}
